src/main/cpp/src/get_json_object.hpp

```diff
@@ -23,6 +23,265 @@
+          if (!is_first_member()) { add_comma(); }
+          // write key
+          write_escaped_text(parser);
+          // add :
+          write_output(":", 1);
+        } else if (next_token == json_token::END_OBJECT) {
+          // add }
+          add_end_object();
+          pop_curr_context();
+        }
+      }
+      // ARRAY Context
+      else {
+        if (next_token == json_token::END_ARRAY) {
+          // add ]
+          add_end_array();
+          pop_curr_context();
+        }
+      }
+    } else {
+      CUDF_FAIL("ERROR:consume_token:INVALID JSON STRUCTURE");
+    }
+    return;
+  }
+
+  CUDF_HOST_DEVICE void consume_value_token(json_token next_token,
+                                            json_parser<max_json_nesting_depth>& parser)
+  {
+    // check if current is a list and add comma if not first member
+    if (!is_context_stack_empty() && is_array_context() && !is_first_member()) { add_comma(); }
+    // make true
+    register_member();
+
+    switch (next_token) {
+      case json_token::START_OBJECT:
+        initialize_new_context(item::OBJECT);
+        add_start_object();
+        break;
+      case json_token::START_ARRAY:
+        initialize_new_context(item::ARRAY);
+        add_start_array();
+        break;
+      case json_token::VALUE_TRUE: add_true(); break;
+      case json_token::VALUE_FALSE: add_false(); break;
+      case json_token::VALUE_NULL: add_null(); break;
+      case json_token::VALUE_NUMBER_INT: write_integer(parser); break;
+      case json_token::VALUE_NUMBER_FLOAT: write_integer(parser); break;
+      case json_token::VALUE_STRING: write_escaped_text(parser); break;
+      default: CUDF_FAIL("ERROR:consume_value_token:INVALID JSON STRUCTURE"); break;
+    }
+  }
+
+  std::string debug_enumToString(json_token token)
+  {
+    switch (token) {
+      case json_token::INIT: return "INIT";
+      case json_token::SUCCESS: return "SUCCESS";
+      case json_token::ERROR: return "ERROR";
+      case json_token::START_OBJECT: return "START_OBJECT";
+      case json_token::END_OBJECT: return "END_OBJECT";
+      case json_token::START_ARRAY: return "START_ARRAY";
+      case json_token::END_ARRAY: return "END_ARRAY";
+      case json_token::FIELD_NAME: return "FIELD_NAME";
+      case json_token::VALUE_STRING: return "VALUE_STRING";
+      case json_token::VALUE_NUMBER_INT: return "VALUE_NUMBER_INT";
+      case json_token::VALUE_NUMBER_FLOAT: return "VALUE_NUMBER_FLOAT";
+      case json_token::VALUE_TRUE: return "VALUE_TRUE";
+      case json_token::VALUE_FALSE: return "VALUE_FALSE";
+      case json_token::VALUE_NULL: return "VALUE_NULL";
+    }
+    return "UNKNOWN_TOKEN";
+  }
+
+  std::string debug_getOutputString()
+  {
+    if (get_output_start_position() != nullptr) {
+      return std::string(get_output_start_position(), get_output_len());
+    } else {
+      return "";
+    }
+  }
+
+ private:
+  bool has_members[max_json_nesting_depth] = {false};
```

Comment:
json_parser is already scanning the json tokens and comma/colon chars.
json_parser also maintains the nesting states.
We can put the writing logic into json_parser when parsing.